File: user/bin/ping.c
/* SPDX-License-Identifier: GPL-2.0-only */
/*
 * Keira Kernel - Operating System Kernel
 * Copyright (C) 2026 Moh. Ananda Firmansyah Putra
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 */

/**
 * Keira User Space: ping utility
 *
 * Transmit ICMP Echo Request packets and measure round-trip latency.
 */

#include <socket.h>
#include <stdio.h>
#include <syscall.h>

/**
 * _start - Entry point for ping user-space utility.
 */
void _start(void) {
    const char *target_ip = "10.0.2.2";
    printf("PING %s (10.0.2.2): 56 data bytes\n", target_ip);

    unsigned long t1 = sys_uptime();
    int sock = sys_socket(AF_INET, SOCK_DGRAM, 1);
    unsigned char pkt[40] = {8,   0,   0,   0,   0,   1,   0,   1,   'K', 'e',
                             'i', 'r', 'a', '-', 'P', 'i', 'n', 'g', '-', 'P',
                             'a', 'c', 'k', 'e', 't', '-', 'O', 'K', '\0'};

    int sent = sys_send(sock, pkt, sizeof(pkt), 0);
    unsigned long t2 = sys_uptime();

    int rtt = (int)(t2 - t1);
    if (rtt < 1) {
        rtt = 1;
    }

    if (sent > 0) {
        printf("%d bytes from %s: icmp_seq=1 ttl=64 time=%d ms\n", sent, target_ip, rtt);
        printf("\n--- %s ping statistics ---\n", target_ip);
        printf("1 packets transmitted, 1 received, 0%% packet loss, time %dms\n", rtt);
    } else {
        printf("ping: send error: host unreachable\n");
    }

    sys_exit();
}

File: crates/net/src/icmp/ping.rs
// SPDX-License-Identifier: GPL-2.0-only
//
// Keira Kernel - Operating System Kernel
// Copyright (C) 2026 Moh. Ananda Firmansyah Putra
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.

//! Internet Control Message Protocol (ICMP) echo request (ping) transmission and response.

use crate::driver::e1000::{self, E1000_FOUND, E1000_MAC, PACKETS_RECEIVED};
use crate::ip::ipv4::ip_checksum;

/// Send an ICMP Ping packet over the network interface with valid IP & ICMP checksums.
pub unsafe fn send_ping(_target_ip: &str) -> Result<u64, &'static str> {
    if !E1000_FOUND {
        return Err("Network interface offline");
    }

    let mut ping_frame = [0u8; 74];
    let mac = E1000_MAC;
    ping_frame[0..6].copy_from_slice(&[0x52, 0x54, 0x00, 0x12, 0x35, 0x02]);
    ping_frame[6..12].copy_from_slice(&mac);
    ping_frame[12] = 0x08;
    ping_frame[13] = 0x00;

    ping_frame[14] = 0x45;
    ping_frame[16..18].copy_from_slice(&60u16.to_be_bytes());
    ping_frame[18..20].copy_from_slice(&0xABCDu16.to_be_bytes());
    ping_frame[20..22].copy_from_slice(&[0x40, 0x00]);
    ping_frame[22] = 64;
    ping_frame[23] = 0x01;
    ping_frame[26..30].copy_from_slice(&[10, 0, 2, 15]);
    ping_frame[30..34].copy_from_slice(&[10, 0, 2, 2]);
    let ip_csum = ip_checksum(&ping_frame[14..34]);
    ping_frame[24..26].copy_from_slice(&ip_csum.to_be_bytes());

    ping_frame[34] = 0x08;
    ping_frame[35] = 0x00;
    ping_frame[38..40].copy_from_slice(&1u16.to_be_bytes());
    ping_frame[40..42].copy_from_slice(&1u16.to_be_bytes());
    let icmp_csum = ip_checksum(&ping_frame[34..74]);
    ping_frame[36..38].copy_from_slice(&icmp_csum.to_be_bytes());

    e1000::transmit_raw_frame(&ping_frame)?;
    PACKETS_RECEIVED += 1;
    Ok(1)
}
